// @iroha2/client@2.0.0/dist/web-socket/node.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var WebSocket = require('ws');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var WebSocket__default = /*#__PURE__*/_interopDefaultLegacy(WebSocket);

function handleIncomingData(data) {
    if (Buffer.isBuffer(data)) {
        return new Uint8Array(data);
    }
    console.error('Bad incoming data:', data);
    throw new Error('Unable to parse incoming data');
}
const adapter = {
    initWebSocket: (params) => {
        const socket = new WebSocket__default["default"](params.url);
        socket.onopen = params.onopen;
        socket.onclose = params.onclose;
        socket.onmessage = (msg) => {
            params.onmessage({
                data: handleIncomingData(msg.data),
            });
        };
        socket.onerror = params.onerror;
        return {
            isClosed: () => socket.readyState === socket.CLOSED,
            send: (data) => socket.send(data),
            close: (code, reason) => socket.close(code, reason),
        };
    },
};

exports.adapter = adapter;
